### PR TITLE
feat(sdk): Introduce `event_cache::Deduplicator`

### PR DESCRIPTION
**Pull request by Hywan**
This patch is based on https://github.com/matrix-org/matrix-rust-sdk/pull/4173 (it must be merged before this one).

This patch should be reviewed commit-by-commit. Only the last two commits are new (the others are part of #4173):

* [2900f69](https://github.com/matrix-org/matrix-rust-sdk/commit/2900f694c833053a0fcbb67b3359d98c4ccc19aa) aims at implementing `Deduplicator`, an efficient type to detect duplicated event. It uses a bloom filter (one we had from our dependencies already), and decorates all events with a `Decoration::Unique` if the event is unique, `Decoration::Duplicated` if it's duplicated, or `Decoration::Invalid` if it's invalid (if it has no `event_id`).
* [8c41d63](https://github.com/matrix-org/matrix-rust-sdk/commit/8c41d63ab92cf4f95ce9f92a442c36f1eac6bc15) uses `Deduplicator` and `LinkedChunk::remove_item_at` (from #4173!) to remove duplicated items from the event cache/the linked chunk.

In the presence of two duplicated events, the older one is removed. It mimics the current behaviour from the `Timeline` API.

### CODE REVIEW

**Comment by Hywan** [comment]
Rebased on `main` since #4173 is now merged.

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+impl Deduplicator {
+    const APPROXIMATED_MAXIMUM_NUMBER_OF_EVENTS: usize = 800_000;
+    const DESIRED_FALSE_POSITIVE_RATE: f64 = 0.001;
+
+    /// Create a new `Deduplicator`.
+    pub fn new() -> Self {
+        Self {
+            bloom_filter: Mutex::new(
+                GrowableBloomBuilder::new()
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/deduplicator.rs:45]: Sorry to be the knight who says [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), but… is it so much faster, considering we do have a btreeset check in both branches after the `check_and_set` below (equivalent to inserting into the btreeset)? Do you have data supporting this (a benchmark)?

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    {
+        let mut already_seen = BTreeSet::new();
+
+        events_to_scan.map(move |event| {
+            let Some(event_id) = event.event_id() else {
+                // The event has no `event_id`.
+                return Decoration::Invalid(event);
+            };
+
+            if self.bloom_filter.lock().unwrap().check_and_set(&event_id) {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/deduplicator.rs:79]: nit: this lock is taken N times where N is the number of events to scan that have event IDs (should be all of them). Can we take the lock above the `.map()` call?

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                // Bloom filter has false positives. We are NOT sure the event is NOT present.
+                // Even if the false positive rate is low, we need to iterate over all events to
+                // ensure it isn't present.
+
+                // But first, let's ensure `event` is not a duplicate from `events`, i.e. if the
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/deduplicator.rs:84]:
```suggestion
                // But first, let's ensure `event` is not a duplicate from `events_to_scan`, i.e. if the
```

File: crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs [hunk 4]
```diff
@@ -948,6 +948,18 @@ impl Position {
     pub fn index(&self) -> usize {
         self.1
     }
+
+    /// Move the index part (see [`Self::index`]) to the left, i.e. subtract 1.
+    ///
+    /// # Panic
+    ///
+    /// This method will panic if it will overflow, i.e. if the index is 0.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/linked_chunk/mod.rs:956]: uber-nit

```suggestion
    /// This method will panic if it will underflow, i.e. if the index is 0.
```

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 5]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+    /// associated event (if it exists) from `Self::chunks`.
+    ///
+    /// This is used to remove duplicated events, see
+    /// [`Self::filter_duplicated_events`].
+    fn remove_events(&mut self, event_ids: Vec<OwnedEventId>) {
+        for event_id in event_ids {
+            let Some(event_position) = self.revents().find_map(|(position, event)| {
+                (event.event_id().as_ref() == Some(&event_id)).then_some(position)
+            }) else {
+                error!(?event_id, "An event has been detected, but it's position seems unknown");
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:219]: The message is slightly too coupled to the deduplication use case IMO, how about something more generic?

```suggestion
                error!(?event_id, "Couldn't find event to remove");
```

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 6]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+    {
+        let mut duplicated_event_ids = Vec::new();
+
+        let deduplicated_events = self
+            .deduplicator
+            .scan_and_learn(events, self)
+            .filter_map(|decorated_event| match decorated_event {
+                Decoration::Unique(event) => Some(event),
+                Decoration::Duplicated(event) => {
+                    error!(event_id = ?event.event_id(), "Found a duplicated event");
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:186]: not an error, it's pretty benign, so `trace` or `debug` please :)

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 7]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+
+                    duplicated_event_ids.push(event.event_id().expect("The event has no ID"));
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:188]: nit: please add safety comment that if it could be marked as invalidated, it must have had an event id in the first place.

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 8]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+
+                    // Keep the new event!
+                    Some(event)
+                }
+                Decoration::Invalid(event) => {
+                    error!(?event, "Found an invalid event");
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:194]:
```suggestion
                    warn!(?event, "Found an event with no event id");
```

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 9]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+                .expect("Failed to remove an event we have just found");
+        }
+    }
+
+    /// Remove all events from `Self::chunks` and update a fix [`Position`].
+    ///
+    /// This method iterates over all event IDs in `event_ids` and removes the
+    /// associated event (if it exists) from `Self::chunks`, exactly like
+    /// [`Self::remove_events`]. The difference is that it will update a
+    /// [`Position`] according to the removals.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:235]: Reading this doc comment, I have no ideas what is `Position` about, how it's been selected, what it represents, sorry. Can you rephrase the doc comment?

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 10]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+    fn remove_events_and_update_position(
+        &mut self,
+        event_ids: Vec<OwnedEventId>,
+        position: &mut Position,
+    ) {
+        for event_id in event_ids {
+            let Some(event_position) = self.revents().find_map(|(position, event)| {
+                (event.event_id().as_ref() == Some(&event_id)).then_some(position)
+            }) else {
+                error!(?event_id, "An event has been detected, but it's position seems unknown");
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:248]:
```suggestion
                error!(?event_id, "Couldn't find event to remove");
```

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 11]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+                .remove_item_at(event_position)
+                .expect("Failed to remove an event we have just found");
+
+            // A `Position` is composed of a `ChunkIdentifier` and an index.
+            // The `ChunkIdentifier` is stable, i.e. it won't change if an
+            // event is removed in another chunk. It means we only need to
+            // update `position` if the removal happened in **the same
+            // chunk**.
+            if event_position.chunk_identifier() == position.chunk_identifier() {
+                // Now we can compare the the position indices.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:263]:
```suggestion
                // Now we can compare the position indices.
```

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 12]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+                        // An event has been removed at the same position of
+                        // the new events: `position` does _NOT_ need tp be
+                        // modified.
+                    }
+
+                    // `event_position`'s index > `position`'s index
+                    Ordering::Greater => {
+                        // An event has been removed _after_ the new events:
+                        // `position` does _NOT_ need to be modified.
+                    }
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:284]:
```suggestion
                    // `event_position`'s index >= `position`'s index
                    Ordering::Equal | Ordering::Greater => {
                        // An event has been removed at the _same_ or _after_ the new events:
                        // `position` does _NOT_ need to be modified.
                    }
```

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 13]
```diff
@@ -352,8 +528,38 @@ mod tests {
...
+            // The first `event_id_0` has been removed.
+            assert_let!(Some((position, event)) = events.next());
+            assert_eq!(position.chunk_identifier(), 0);
+            assert_eq!(position.index(), 0);
             assert_eq!(event.event_id().unwrap(), event_id_1);
 
+            assert_let!(Some((position, event)) = events.next());
+            assert_eq!(position.chunk_identifier(), 0);
+            assert_eq!(position.index(), 1);
+            assert_eq!(event.event_id().unwrap(), event_id_0);
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:551]: Could we have a small assert helper function (if not a `macro_rules!`) that checks chunk_identifier, index and event id?

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 14]
```diff
@@ -519,4 +744,257 @@ mod tests {
...
+
+                assert_let!(Some((position, event)) = events.next());
+                assert_eq!(position.chunk_identifier(), 2);
+                assert_eq!(position.index(), 0);
+                assert_eq!(event.event_id().unwrap(), event_id_7);
+
+                assert_let!(Some((position, event)) = events.next());
+                assert_eq!(position.chunk_identifier(), 2);
+                assert_eq!(position.index(), 1);
+                assert_eq!(event.event_id().unwrap(), event_id_8);
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:901]: ouch yeah, let's introduce a test helper for that

File: crates/matrix-sdk/src/event_cache/store.rs [hunk 15]
```diff
@@ -130,6 +166,126 @@ impl RoomEvents {
...
+    /// Remove all events from `Self::chunks` and update a fix [`Position`].
+    ///
+    /// This method iterates over all event IDs in `event_ids` and removes the
+    /// associated event (if it exists) from `Self::chunks`, exactly like
+    /// [`Self::remove_events`]. The difference is that it will update a
+    /// [`Position`] according to the removals.
+    ///
+    /// This is used to remove duplicated events, see
+    /// [`Self::filter_duplicated_events`].
+    fn remove_events_and_update_position(
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/store.rs:239]: Now that I understand what it does: how about renaming `remove_events_and_update_insert_position`?

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+                // iterator itself contains duplicated events! We use a `BTreetSet`, otherwise
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/deduplicator.rs:85]: `BTroutSet` :fish: 

```suggestion
                // iterator itself contains duplicated events! We use a `BTreeSet`, otherwise
```

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+    /// Each scanned event will update `Self`'s internal state.
+    ///
+    /// `existing_events` represents all events of a room that already exist.
+    pub fn scan_and_learn<'a, I>(
+        &'a self,
+        events_to_scan: I,
+        existing_events: &'a RoomEvents,
+    ) -> impl Iterator<Item = Decoration<I::Item>> + 'a
+    where
+        I: Iterator<Item = Event> + 'a,
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/deduplicator.rs:69]: I'm not sure about this API's signature: instead of taking an iterator and returning one, could we have a function that operates on a single item at a time, returning a decoration for it? `already_seen` could be a member of this struct (which would be more optimized too), and with respect to the YAGNI about the bloom filter, we could likely only use `already_seen`…

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 18]
```diff
@@ -0,0 +1,266 @@
...
+        SyncTimelineEvent::new(event_builder.make_sync_message_event_with_id(
+            *ALICE,
+            event_id,
+            RoomMessageEventContent::text_plain("foo"),
+        ))
+    }
+
+    #[test]
+    fn test_filter_no_duplicate() {
+        let event_builder = EventBuilder::new();
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs:148]: Good catch. Done!

File: crates/matrix-sdk/src/event_cache/deduplicator.rs (hunk 1 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs:45]: The `BTreeSet` is used only to check the duplicates in the _new_ events that we are checking. The bloom filter is to check the duplicates in _past_ events. For example:

* The bloom filter contains `[]`. I check `["$ev0", "$ev1", "$ev2]`. No duplicates.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2"]`. I check `["$ev3"]`. No duplicates.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2", "$ev3"]`. I check `["$ev0"]`. There is a duplicate.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2", "$ev3"]`. I check `["$ev4", "$ev4"]`. There is a duplicate but in the _new_ events! That's detected by the use of `BTreeSet`.
* The bloom filter now contains `["$ev0", "$ev1", "$ev2", "$ev3", "$ev4"]` now.

This `BTreeSet` is necessary to efficiently detect duplicates in the _new_ scanned events. If we rely on the Bloom filter only, we cannot check the false positives, because the _new_ events are not yet added into the `RoomEvents`.

You're right the `BTreeSet` is hit in the 2 paths inside `check_and_set` but the bloom filter is still more memory-space efficient than if we were using `BTreeSet` for _new_ (currently scanned) **and** _past_ (already scanned) events. That's the only reason I use a bloom filter here: to save memory.

File: crates/matrix-sdk/src/event_cache/deduplicator.rs (hunk 2 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs:79]: Sadly no because the iterator is returned, so it would outlive the lock…

File: crates/matrix-sdk/src/event_cache/deduplicator.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs:84]: It's indeed a typo, but it's `existing_events`, not `events_to_scan`.

Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs:84]: Thanks!

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 5 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:219]: I agree. I was hesitant keeping this message. Updating it!

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 6 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:186]: Went for a `debug!`.

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 7 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:188]: Okay.

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 8 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:194]: Better :+1:.

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 11 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:263]: Done.

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 12 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:284]: Good idea.

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 13 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:551]: Sir, yes Sir!

```diff
-            {
-                let mut events = room_events.events();
-
-                assert_let!(Some((position, event)) = events.next());
-                assert_eq!(position.chunk_identifier(), 0);
-                assert_eq!(position.index(), 0);
-                assert_eq!(event.event_id().unwrap(), event_id_1);
-
-                assert_let!(Some((position, event)) = events.next());
-                assert_eq!(position.chunk_identifier(), 0);
-                assert_eq!(position.index(), 1);
-                assert_eq!(event.event_id().unwrap(), event_id_2);
-
-                assert_let!(Some((position, event)) = events.next());
-                assert_eq!(position.chunk_identifier(), 0);
-                assert_eq!(position.index(), 2);
-                assert_eq!(event.event_id().unwrap(), event_id_3);
-
-                assert_let!(Some((position, event)) = events.next());
-                assert_eq!(position.chunk_identifier(), 0);
-                assert_eq!(position.index(), 3);
-                assert_eq!(event.event_id().unwrap(), event_id_6);
-
-                assert_let!(Some((position, event)) = events.next());
-                assert_eq!(position.chunk_identifier(), 2);
-                assert_eq!(position.index(), 0);
-                assert_eq!(event.event_id().unwrap(), event_id_7);
-
-                assert_let!(Some((position, event)) = events.next());
-                assert_eq!(position.chunk_identifier(), 2);
-                assert_eq!(position.index(), 1);
-                assert_eq!(event.event_id().unwrap(), event_id_8);
-
-                assert!(events.next().is_none());
-            }
+            assert_events_eq!(
+                room_events.events(),
+                [
+                    (event_id_1 at (0, 0)),
+                    (event_id_2 at (0, 1)),
+                    (event_id_3 at (0, 2)),
+                    (event_id_6 at (0, 3)),
+                    (event_id_7 at (2, 0)),
+                    (event_id_8 at (2, 1)),
+                ]
+            );
```

It's pretty useful :-).

File: crates/matrix-sdk/src/event_cache/deduplicator.rs (hunk 16 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs:85]: <3

File: crates/matrix-sdk/src/event_cache/store.rs (hunk 15 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/store.rs:239]: Okay with the name.

File: crates/matrix-sdk/src/event_cache/deduplicator.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs]: …Not done?

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+
+    #[test]
+    fn test_filter_no_duplicate() {
+        let event_id_0 = owned_event_id!("$ev0");
+        let event_id_1 = owned_event_id!("$ev1");
+        let event_id_2 = owned_event_id!("$ev2");
+
+        let event_0 = sync_timeline_event(&event_id_0);
+        let event_1 = sync_timeline_event(&event_id_1);
+        let event_2 = sync_timeline_event(&event_id_2);
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/deduplicator.rs:160]: Time for an `EventFactory 101` :nerd_face: You can curry some parameters at the factory level, to avoid dull repetition, which would justify *not* having to have the `sync_timeline_event` helper:

```suggestion
        let f = EventFactory::new().sender(user_id!("@mnt_io:matrix.org"));
        let event_0 = f.text_msg("").event_id(&event_id_0).into_sync();
        let event_1 = f.text_msg("").event_id(&event_id_1).into_sync();
        let event_2 = f.text_msg("").event_id(&event_id_2).into_sync();
```

File: crates/matrix-sdk/src/event_cache/room/events.rs [hunk 20]
```diff
@@ -129,24 +167,182 @@ impl RoomEvents {
...
+    fn remove_events_and_update_insert_position(
+        &mut self,
+        event_ids: Vec<OwnedEventId>,
+        position: &mut Position,
+    ) {
+        for event_id in event_ids {
+            let Some(event_position) = self.revents().find_map(|(position, event)| {
+                (event.event_id().as_ref() == Some(&event_id)).then_some(position)
+            }) else {
+                error!(?event_id, "Cannot find te event to remove");
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/room/events.rs:277]:
```suggestion
                error!(?event_id, "Cannot find the event to remove");
```

File: crates/matrix-sdk/src/event_cache/room/events.rs [hunk 21]
```diff
@@ -129,24 +167,182 @@ impl RoomEvents {
...
+                        // events: `position` needs to be shifted to the
+                        // left by 1.
+                        position.move_index_to_the_left();
+                    }
+
+                    // `event_position`'s index >= `position`'s index
+                    Ordering::Equal | Ordering::Greater => {
+                        // An event has been removed at the _same_ position of
+                        // or _after_ the new events: `position` does _NOT_ need
+                        // tp be modified.
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/room/events.rs:306]:
```suggestion
                        // to be modified.
```

File: crates/matrix-sdk/src/event_cache/room/events.rs [hunk 22]
```diff
@@ -129,24 +167,182 @@ impl RoomEvents {
...
+            // chunk**.
+            if event_position.chunk_identifier() == position.chunk_identifier() {
+                // Now we can compare the position indices.
+                match event_position.index().cmp(&position.index()) {
+                    // `event_position`'s index < `position`'s index
+                    Ordering::Less => {
+                        // An event has been removed _before_ the new
+                        // events: `position` needs to be shifted to the
+                        // left by 1.
+                        position.move_index_to_the_left();
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/room/events.rs:299]: Could `move_index_to_the_left` be called `decrement_index`?

File: crates/matrix-sdk/src/event_cache/room/events.rs [hunk 23]
```diff
@@ -129,24 +167,182 @@ impl RoomEvents {
...
+                let mut events = $events_iterator;
+
+                $(
+                    assert_let!(Some((position, event)) = events.next());
+                    assert_eq!(position.chunk_identifier(), $chunk_identifier );
+                    assert_eq!(position.index(), $index );
+                    assert_eq!(event.event_id().unwrap(), $event_id );
+                )*
+
+                assert!(events.next().is_none(), "No more event is expected");
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/room/events.rs:334]: Zero is plural in English :nerd_face: 

```suggestion
                assert!(events.next().is_none(), "No more events are expected");
```

File: crates/matrix-sdk/src/event_cache/room/events.rs [hunk 24]
```diff
@@ -129,24 +167,182 @@ impl RoomEvents {
...
 mod tests {
     use assert_matches2::assert_let;
-    use matrix_sdk_test::{EventBuilder, ALICE};
-    use ruma::{events::room::message::RoomMessageEventContent, EventId, OwnedEventId};
+    use ruma::{user_id, EventId, OwnedEventId};
 
     use super::*;
+    use crate::test_utils::events::EventFactory;
+
+    macro_rules! assert_events_eq {
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/room/events.rs:322]: So much nicer with this macro :star_struck:

File: crates/matrix-sdk/src/event_cache/deduplicator.rs (hunk 19 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs:160]: Thanks for the tips.

I agree it's nice but in this case it adds a non-negligible cognitive effort to read `.text_msg("")` and `.into_sync();` here. I think it's fine to keep `sync_timeline_event` as it makes the test code easier to understand, and we can focus on the real things we want to test rather than the prelude.

File: crates/matrix-sdk/src/event_cache/deduplicator.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/deduplicator.rs]: I don't use `EventBuilder` but I still prefer to use a helper to make the prelude of the test easier simpler. See my comment here https://github.com/matrix-org/matrix-rust-sdk/pull/4174#discussion_r1820902651.

File: crates/matrix-sdk/src/event_cache/room/events.rs (hunk 20 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/room/events.rs]: 🤦  thanks!

File: crates/matrix-sdk/src/event_cache/room/events.rs (hunk 22 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/room/events.rs]: Sure.

File: crates/matrix-sdk/src/event_cache/room/events.rs (hunk 21 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/room/events.rs]: 🤦

File: crates/matrix-sdk/src/event_cache/room/events.rs (hunk 24 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/room/events.rs:322]: It is!

File: crates/matrix-sdk/src/event_cache/room/events.rs (hunk 23 above)
Comment (Hywan) [reply crates/matrix-sdk/src/event_cache/room/events.rs]: TIL

File: crates/matrix-sdk/src/event_cache/deduplicator.rs [hunk 25]
```diff
@@ -0,0 +1,266 @@
...
+    /// `existing_events` represents all events of a room that already exist.
+    pub fn scan_and_learn<'a, I>(
+        &'a self,
+        events_to_scan: I,
+        existing_events: &'a RoomEvents,
+    ) -> impl Iterator<Item = Decoration<I::Item>> + 'a
+    where
+        I: Iterator<Item = Event> + 'a,
+    {
+        let mut already_seen = BTreeSet::new();
```
Comment (bnjbvr) [inline crates/matrix-sdk/src/event_cache/deduplicator.rs:77]: 1. Can you comment that this is a temporary variable (and not stored in `Deduplicator`) because we don't want an unbounded growth for this set? (and that it justifies the shape of the API, taking an in-parameter that's an iterator)
2. Can you also comment what its purpose is, here (i.e. find duplicates in the incoming iterator), please?

**Review by bnjbvr — APPROVED** [review]
